internal/logsink/logsink.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package logsink
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/juju/loggo/v2"
+	"gopkg.in/tomb.v2"
+)
+
+const (
+	stateFlushed = "flushed"
+	stateTicked  = "ticked"
+)
+
+var (
+	zeroTime = time.Time{}
+)
+
+// NewWriterFunc is a function that creates a new writer.
+type NewWriterFunc func() (io.WriteCloser, error)
+
+// LogSink is a loggo.Writer that writes log messages to a file.
+type LogSink struct {
+	tomb           tomb.Tomb
+	internalStates chan string
+
+	writer io.Writer
+
+	batchSize     int
+	flushInterval time.Duration
+
+	in  chan loggo.Entry
+	out chan []logRecord
+
+	pool sync.Pool
+}
+
+// NewLogSink creates a new log sink that writes log messages to a file. There
+// can only be one writer writing to the same file at a time, otherwise bytes
+// will be written to the file in an interleaved manner (junk data).
+// LogSink writer will write log messages as JSON objects, one per line, even
+// if the log message is multiline.
+func NewLogSink(writer io.Writer, batchSize int, flushInterval time.Duration) *LogSink {
+	return newLogSink(writer, batchSize, flushInterval, nil)
+}
+
+// newLogSink creates a new log sink that writes log messages to a file.
+func newLogSink(writer io.Writer, batchSize int, flushInterval time.Duration, internalStates chan string) *LogSink {
+	w := &LogSink{
+		internalStates: internalStates,
+
+		writer: writer,
+
+		batchSize:     batchSize,
+		flushInterval: flushInterval,
+
+		in:  make(chan loggo.Entry, batchSize),
+		out: make(chan []logRecord, batchSize),
```

Comment:
The `batchSize` does not make sense for the `out` channel (since the log records should be batched already). I'd recommend a different tuneable for the `out` channel.